File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque

x,y,z = map(int, input().split())
graph = [ [0]*(x+1) for _ in range(x+1)]
visited = [0] * (x+1)
arr = []
for i in range(y):
    a, b = map(int,input().split())
    graph[a][b] = graph[b][a] = 1

def dfs(graph, visited, start):
    visited[start] = 1
    print(start, end=' ')

    for i in range(1,x+1):
        if graph[start][i] == 1 and visited[i] == 0:
            visited[i] = 1
            dfs(graph,visited,i)
    

def bfs(graph, visited, start):
    q = deque([start])
    visited[start] = 1

    while q:
        v = q.popleft()
        print(v, end=' ')
        for i in range(1,x+1):
            if graph[v][i] == 1 and visited[i] == 0:
                q.append(i)
                visited[i] = 1
                
dfs(graph, visited, z)
print("")

visited = [0] * (x+1)
bfs(graph,visited,z)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;

public class Main {
	
	//---------------------dfs-------------------
	static void DFS(int start,int N, int[][] graph,boolean[] visited,List<Integer> arr) {
		for (int k =1; k<N+1;k++) {
			if (graph[start][k] == 1 && visited[k] == false) {
				visited[k] = true;
				arr.add(k);
				DFS(k,N,graph,visited,arr);
			}
		}
	}
	

	public static void main(String[] args) throws IOException {
		
		//입력
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int V = Integer.parseInt(st.nextToken());
		
		int[][] graph = new int[N+1][N+1];
		boolean[] visited = new boolean[N+1];
		List<Integer> arr = new LinkedList<>();
		
		
		for (int i = 0; i < M; i ++) {
			StringTokenizer st2 = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st2.nextToken());
			int b = Integer.parseInt(st2.nextToken());
			
			graph[a][b] = graph[b][a] = 1;
		}
		
		//---------------------dfs-------------------
		visited[V] = true;
		arr.add(V);
		DFS(V,N,graph,visited,arr);
		
		//---------------------bfs-------------------
		
		boolean[] visited2 = new boolean[N+1];
		List<Integer> arr2 = new LinkedList<>();
		
		//판별
		Queue<Integer> queue = new LinkedList<>(); 
		queue.add(V);
		arr2.add(V);
		visited2[V] = true;
		
		while(!queue.isEmpty()) {
			int current = queue.poll();
			
			//숫자가 작은 순으로 방문
			for(int j = 1; j<N+1; j++) {
				if (graph[current][j] ==1 && visited2[j] ==false) {
					visited2[j] = true;
					queue.add(j);
					arr2.add(j);
				}
			}
		}
		//dfs 출력
		for(int l = 0; l<arr.size(); l++) {
			System.out.print(arr.get(l)+" ");
		}
		System.out.println();
		//bfs 출력
		for(int l = 0; l<arr.size(); l++) {
			System.out.print(arr2.get(l)+" ");
		}
	}
}
